commands/utility/warn.js

```diff
@@ -0,0 +1,127 @@
+}
+
+function recordInDB(msg,con,offendingUser,warningReason){
+	// Add the infraction to the database
+	let now = new Date();
+	let timestamp = dateFormat(now, "yyyy-mm-dd HH:MM:ss");
+
+	const sqlInfractions = `INSERT INTO infractions (timestamp, user, action, length_of_time, reason, valid, moderator) VALUES ('${timestamp}', '${offendingUser.id}', 'cc!warn', NULL, '${warningReason}', true, '${msg.author.id}')`;
+	const sqlModLog = `INSERT INTO mod_log (timestamp, moderator, action, length_of_time, reason) VALUES ('${timestamp}', '${msg.author.id}', '${msg}', NULL, '${warningReason}')`;
+
+	con.query(`${sqlInfractions}; ${sqlModLog}`, function (err, result) {
+		if (err) {
+			console.log(err);
+			// Include a warning in case something goes wrong writing to the db
+			msg.channel.send(`I warned ${offendingUser} but writing to the db failed!`);
+		} else {
+			console.log("1 record inserted into infractions, 1 record inserted into mod_log.");
+		}
+	});
+}
+
+function canWarn(msg) {
+	if (!msg.member.roles.cache.some(
+		role => role.name === "Super User" || role.name === "Moderator" || role.name === "Admin")) {
+		msg.reply("You must be a Super User, Moderator or Admin to use this command.");
+		return false;
+	} else {
+		return true;
+	}
+}
+
+function hasUserTarget(msg) {
+	// Asortment of answers to make the bot more fun
+	const failAttemptReply = [
+		"Ok there bud, who are you trying to warn again?",
+		"You definitely missed the targer user there...",
+		"shoot first ask later? You forgot the targer user",
+		"Not judging, but you didn't set a user to warn",
+		"Without a target user I can't warn anyone but you",
+		"Forgot the target user. Wanna try again?",
+		"Please tell you *do* know who to warn? You forgot the user"
+		];
+
+	const offendingUser = msg.mentions.members.first();
+	if (offendingUser){
+		return true;
+	} else {
+		msg.reply(failAttemptReply[Math.floor(Math.random() * failAttemptReply.length)]);
+		return false;
+	}
+}
+
+function notHighRoller (msg){
+	const offendingUser = msg.mentions.members.first();
+	if (offendingUser.roles.cache.some(
+		role => role.name === "Super User" || role.name === "Moderator" || role.name === "Admin")) {
+			msg.reply("You cannot warn a super user, moderator or admin.");
+			return false;
+	} else {
+		return true;
+	}
+}
+
+function notSelf(msg) {
+	const offendingUser = msg.mentions.members.first();
+	if (offendingUser.id == msg.author.id){
+		msg.reply('Did you just try to warn yourself?');
+		return false;
+	} else {
+		return true;
+	}
+}
+
+module.exports = {
+	name: 'warn',
+	description: 'warns a user of an infraction and logs infraction in db',
+	execute(msg,con,args) {
+		// Make sure only SU, Mods and Admin can run the command
+		if (canWarn(msg)){
+			if (hasUserTarget(msg) && notHighRoller(msg) && notSelf(msg)) {
+				// Parse the reason for the warning
+				// if no reason provided, return so the bot doesn't go boom
+				const warningReason = args.slice(1).join(" ");
```

Comment:
Just two things: 

First, I think we want to put `&& notSelf` before `&& notHighRoller` so we can get the warning about ourselves if we are also admin/mod/SU.  Right now, if I try to warn myself I get the message about the admin.

Second, I think the best place to define `offendingUser` is on line 107 right after `if (canWarn(msg))`. All three of the functions in the line 107 conditional re-define `offendingUser`, so it would make sense to just pass it in as a param:

```javascript
const offendingUser = msg.mentions.members.first();
if ((hasUserTarget(msg, offendingUser) && notSelf(msg, offendingUser) && notHighRoller(msg, offendingUser))
```
that way, it's in the execute function and it's only defined the one time.

Other than that, the code looks good to me!

Reply:
Both things covered